### PR TITLE
Add dictionary_expresions feature (#4386)

### LINKED ISSUE #4386: Make Binary Dictionary Operations Optional
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

The dyn_cmp_dict and dyn_arith_dict features of arrow enable binary operations involving dictionary arrays and other dictionary arrays or scalar arrays.

They are, however, extremely expensive both from a compilation time, and code size perspective. As the kernels must be generated for the combinatorial explosion of all dictionary key and value types. See https://github.com/apache/arrow-rs/issues/2596 and https://github.com/apache/arrow-rs/pull/2760.

They are also exceedingly rare in practice, as almost all queries instead use the scalar variant, i.e. add scalar value to dictionary, compare dictionary against scalar value, etc...

**Describe the solution you'd like**

I would like a feature flag, e.g. binary_dict_op, that is not enabled by default, and enables the arrow features. The three or so tests that happen to need this, can then be gated on this feature flag or possibly updated to not need it.

**Describe alternatives you've considered**

We could not do this

**Additional context**

The feature was initially enabled as a dev dependency in #3363 by @avantgardnerio 

This was then updated as physical-expr dev dependency in #4163 by @isidentical 

It was then enabled as a non-dev dependency in #4168 by @retikulum possibly unintentionally?

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #4386

# Rationale for this change

Release build of datafusion-cli on master

```
________________________________________________________
Executed in   65.44 secs    fish           external
   usr time   17.23 mins  521.00 micros   17.23 mins
   sys time    0.48 mins    0.00 micros    0.48 mins
```

With this feature

```
________________________________________________________
Executed in   48.18 secs    fish           external
   usr time   16.46 mins  567.00 micros   16.46 mins
   sys time    0.48 mins    0.00 micros    0.48 mins

```

# What changes are included in this PR?

Adds a `dictionary_expressions` feature that gates `arrow/dyn_cmp_dict` and `arrow/dyn_arith_dict`. I vacillated a bit on the naming and whether this should enable `arrow/dyn_arith_dict` but figured as this is just plumbing features through, and does not alter the actual DataFusion code, downstreams could not use this feature and manually enabled the desired `arrow` features themselves if they so wish.

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/core/tests/path_partition.rs [hunk 1]
```diff
@@ -204,7 +204,7 @@ async fn csv_filter_with_file_col() -> Result<()> {
     );
 
     let result = ctx
-        .sql("SELECT c1, c2 FROM t WHERE date='2021-10-27' and date!=c1 LIMIT 5")
+        .sql("SELECT c1, c2 FROM t WHERE date='2021-10-27' and c1!='2021-10-27' LIMIT 5")
```
Comment (tustvold) [inline datafusion/core/tests/path_partition.rs:207]: I couldn't see a compelling reason why this test needed to test comparison of dictionary columns

File: datafusion/physical-expr/Cargo.toml [hunk 2]
```diff
@@ -35,12 +35,15 @@ path = "src/lib.rs"
 [features]
 crypto_expressions = ["md-5", "sha2", "blake2", "blake3"]
 default = ["crypto_expressions", "regex_expressions", "unicode_expressions"]
```
Comment (alamb) [inline datafusion/physical-expr/Cargo.toml:37]: I think we should keep dictionary support as a default, if possible

**Review by alamb — APPROVED** [review]
Seems reasonable to me -- it would be nice to have dictionary support enabled by default, but not required

File: datafusion/physical-expr/Cargo.toml (hunk 2 above)
Comment (tustvold) [reply datafusion/physical-expr/Cargo.toml:37]: I fairly strongly disagree, it is pretty esoteric. As a data point, none of IOx's integration tests require this, and we use dictionaries a LOT :smile:

Comment (alamb) [reply datafusion/physical-expr/Cargo.toml:37]: > I fairly strongly disagree, it is pretty esoteric. 

I am not sure how to quantify how esoteric the feature is or how commonly it is used. Clearly IOx uses it. I was just thinking that this PR changes the default behavior

But maybe that is ok.

Perhaps some other committers have thoughts.

Comment (tustvold) [reply datafusion/physical-expr/Cargo.toml:37]: > Clearly IOx uses it

That is the point I'm trying to make, IOx **doesn't** use it, at least not within any of its tests

Comment (alamb) [reply datafusion/physical-expr/Cargo.toml:37]: > at least not within any of its tests.

🤔  when I didn't enable the dyn dictionary kernels in arrow iox tests failed in some past version

We have it enabled here: https://github.com/influxdata/influxdb_iox/blob/6f39ae342e64848bd6555bddbc1d3fa30050f75e/arrow_util/Cargo.toml#L12

File: datafusion/core/tests/path_partition.rs (hunk 1 above)
Comment (alamb) [reply datafusion/core/tests/path_partition.rs:207]: I think the partitioning columns in a ListingTable are dictionary encoded

**Pull request merged**